### DISCUSSION TITLE
More than one foreign key from the same table

### DISCUSSION THREAD

**Discussion by aymanalqadhi** [diesel-rs/diesel#2974] [Q&A] [answered]
Hey all,

So I have this ERD that I want to map using diesel:
![image](https://user-images.githubusercontent.com/39128795/145691210-033762b3-d0a0-4496-af75-cfb17d433ee8.png)

I want to map the entity "Transaction" using rust. The SQL I ended up for the entity is as the following:
```sql
CREATE TABLE transactions (
        id          SERIAL    PRIMARY KEY,
        amount      NUMERIC   NOT NULL,
        state       BOOLEAN   NULL,
        signature   VARCHAR   NOT NULL,

        sender      SERIAL    NOT NULL REFERENCES users(id),
        receiver    SERIAL    NOT NULL REFERENCES users(id),
        made_at     TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT CURRENT_TIMESTAMP,

        verified_by SERIAL    REFERENCES users(id)
)
```

The model code that I wrote to map the database table is as follows:
```rust
#[derive(Associations, Clone, Debug, Identifiable, Queryable, Validate)]
#[belongs_to(User, foreign_key = "sender")]
#[belongs_to(User, foreign_key = "receiver")]
pub struct Transaction {
    pub id: KeyType,

    pub amount: BigDecimal,
    pub state: Option<bool>,
    pub signature: String,

    pub sender: KeyType,
    pub receiver: KeyType,
    pub made_at: chrono::DateTime<chrono::Utc>,
}
```

When I try to compile, I get this error:
```rust
error[E0119]: conflicting implementations of trait `diesel::associations::BelongsTo<data::models::user::User>` for type `data::models::financial_record::Transaction`
  --> wasl/src/data/models/financial_record.rs:23:10
   |
23 | #[derive(Associations, Clone, Debug, Identifiable, Queryable, Validate)]
   |          ^^^^^^^^^^^^
   |          |
   |          first implementation here
   |          conflicting implementation for `data::models::financial_record::Transaction`
   |
   = note: this error originates in the derive macro `Associations` (in Nightly builds, run with -Z macro-backtrace for more info)
```

I have done some research, and found that diesel does not support multiple uses of the `#[belongs_to]` attribute because `BelongsTo` trait can only be implemented once per struct.
Is there any way to solve this problem? or if not, how can I map the entity without arriving to this situation?

**Comment by weiznich** [comment, marked as answer]
Let me start with some reasoning why multiple `belongs_to` attribute referencing the same table are not allowed. The issue there is that we cannot say which of the referenced dependencies should be used to resolve `Transactions::belonging_to(&user)`. We could likely require that you always need to specify the used foreign key column, but that would add unneeded boiler plate code for the case where only on `belongs_to` attribute is given. 

So how to "fix" this: As `Transactions::belonging_to(&user)` only constructs a simple query (`users::table.filter(users::id.eq_any(foreign_keys))`) you can just write this query explicitly. 

What could be improved here is the derive itself, so that we already issue an error there indicating this solution. If someone is interested in contributing this improvement I can add details about the potential implementation.

**Reply by Mingun** [reply]
As alternative I suggest to add a new method `Transactions::belonging_to_by(&user, <column>)` with the second parameter that allows you to specify what column should be used.

**Reply by weiznich** [reply]
@Mingun Please just stop to make suggestions that just repeat what I've written above, without trying to even address the reasons why we did not implement that. Such suggestions are not welcome here if that's your "only" interaction with the project.

**Reply by Mingun** [reply]
You written the different thing if my eyes not lies me. At least I do not see how you suggestion to just produce a compiler error for @aymanalqadhi use-case will help him (please do not tell me that he can write everything by hands. Task NOT to write this manually).

At the same time the solution lies on the surface. If you cannot decide automatically what the user want you always can require a parameter from it to make a decision.

<details>
Sorry, but for now all I see is you're trying to turn a blind eye to users' problems without real alternative.
</details>

**Reply by aymanalqadhi** [reply]
@weiznich
Why would the solution that @Mingun suggested be a bad one?

**Reply by weiznich** [reply]
> Why would the solution that @Mingun suggested be a bad one?

Because that's not possible to implement, at least not by "just add another method or attribute." 

Let me explain that a bit more in detail:

`Transactions::belonging_to` uses [`BelongingToDsl`](https://docs.diesel.rs/1.4.x/diesel/prelude/trait.BelongingToDsl.html). Just adding a `belogning_to_by` method there is not possible as we would need to have a type for the foreign key column. @Minigun just let of that information from their "suggestion". I would such call "simplifications" common for their comments on this repo. So why is this a problem at all: Each column has a unique type, therefore we need some way to specify that type as part of the trait definition. `BeloningToDsl` uses the definition by [`BelongsTo`](https://docs.diesel.rs/1.4.x/diesel/associations/trait.BelongsTo.html). This is the trait that `#[derive(Associations]]` implements for your type. As you can see there this trait defines the type of the foreign key column as associated type. Now you could say: Lets just use this type and I would answer that this is possible but would not solve your problem. In fact you see the problem as part of the error message above: You cannot implement the same trait twice for the same struct if all generic parameters are the same, even if any associated type is different. Thats a language restriction, so that's not up to discussion. 
Now you could argue: Lets just change the trait to include the foreign key as well. Again thats something that is possible and would likely solve your problem. So whats my problem with that solution: There wouldn't be any way to differentiate between associations with one foreign key relation and with those more than one relation. The problem with that is that it would very likely require that the foreign key is always required as part of the `Transactions::beloging_to(&user)` call, even if there is only one matching relation. Thats something I find not a acceptable change, especially as the alternative contains basically the same fields and is not any longer to write: `transactions.filter(foreign_key.eq(user.id))`.
Now one could argue that there must be an other solution here and that might be true. I'm happy to see an concrete proposal here but at least for me its not something with priority (as my use cases do not have many of those case) and I want to spend my limited time to work something else more important for my own use cases.

@Mingun To repeat that again: Your repeated simplified "suggestions" how to solve problems are not welcome here. Either spend the time to do at least a minimal amount of checking on how to implement your suggestions or just stop posting them. Otherwise I must consider your posts as hostile.

**Reply by aymanalqadhi** [reply]
Thanks for your answer. I guess I will use your alternative solution for the moment.

**Reply by Mingun** [reply]
> Because that's not possible to implement, at least not by "just add another method or attribute."

@weiznich , you said absolutely the same thing for the `Selectable` trait, so let me not believe you here either. And of course you should add some code that allow you to write that method -- this is obvious.

Please cool down. I'm just telling you that your opinion may not be the only right one. Since you did not consider the most obvious option in your decision, I pointed you to it, suddenly you missed it. How to implement it is a task for maintainers. I am not interested in writing code for diesel at this time.

------
> Just adding a `belonging_to_by` method there is not possible as we would need to have a type for the foreign key column.

Did I say that you should add the method necessarily to this trait and nowhere else? Make a new trait if necessary. This is absolutely unimportant at this stage. Just _think_ about how ерш can be solved using this function and design solution accordingly.

> You cannot implement the same trait twice for the same struct if all generic parameters are the same, even if any associated type is different.

You are attached to the existing code and do not want to look further. What makes you want to use `BelongsTo`? There is some fundamental limitation, why is it impossible to introduce a new trait? Something like:
```rust
pub trait BelongsToBy<Parent, ForeignKeyColumn: Column> {
   ...
}
```
As I see, you find this solution by yourself.

> There wouldn't be any way to differentiate between associations with one foreign key relation and with those more than one relation.

And we need this? For what? This looks like you trying to solve a problem that even not exists.

> The problem with that is that it would very likely require that the foreign key is always required as part of the `Transactions::belonging_to(&user)` call, even if there is only one matching relation.

Not required, if `belonging_to` would use some selected column (default relation) or only be implemented for tables with one relation (using derive, you still could implement it manually and pick up some default column to use in that case).

> I want to spend my limited time to work something else more important for my own use cases.

You say as if I demand that you quit your work and immediately realize this feature. It is not true. Unless this is your personal hobby project (mention of which I did not find in readme) it is quite natural to invite the _maintainers_ to think about what they will do to _maintain_ the project.

**Reply by weiznich** [reply]
@Mingun I do not want to repeat my arguments here. I mark your comment as off topic now as it does not answer the question of the user. Suggestions about potential improvements are welcome, but not as "drive by" comment on questions. If you believe this feature could be improved please open a new topic [here](https://github.com/diesel-rs/diesel/discussions/categories/ideas) outlining the potential improvement. This should include at least some details on how this could be implemented and how this would affect the public API of diesel. That written: I'm not be able to spend any more time on discussing those half finished suggestions with you. Either spend the time to write a concrete proposal or stop posting them. And please don't tell me how to spend my time maintaining diesel.
